llama-index-integrations/vector_stores/llama-index-vector-stores-redis/llama_index/vector_stores/redis/base.py

```diff
@@ -157,41 +216,47 @@ def add(self, nodes: List[BaseNode], **add_kwargs: Any) -> List[str]:
         Raises:
             ValueError: If the index already exists and overwrite is False.
         """
-        # check to see if empty document list was passed
+        # Check to see if empty document list was passed
         if len(nodes) == 0:
             return []
 
-        # set vector dim for creation if index doesn't exist
-        self._index_args["dims"] = len(nodes[0].get_embedding())
+        # Now check for the scenario where user is trying to index embeddings that don't align with schema
+        embedding_len = len(nodes[0].get_embedding())
+        expected_dims = self._index.schema.fields[VECTOR_FIELD_NAME].attrs.dims
+        if expected_dims != embedding_len:
+            raise ValueError(
+                f"Attempting to index embeddings of dim {embedding_len} "
+                f"which doesn't match the index schema expectation of {expected_dims}. "
+                "Please review the Redis integration example to learn how to customize schema."
+            )
 
-        if self._index_exists():
-            if self._overwrite:
-                self.delete_index()
-                self._create_index()
-            else:
-                logging.info(f"Adding document to existing index {self._index_name}")
+        # Create index honoring overwrite policy
+        if self._overwrite:
+            self._index.create(overwrite=True, drop=True)
         else:
-            self._create_index()
+            self._index.create()
 
-        ids = []
+        data: List[Dict[str, Any]] = []
         for node in nodes:
-            mapping = {
-                "id": node.node_id,
-                "doc_id": node.ref_doc_id,
-                "text": node.get_content(metadata_mode=MetadataMode.NONE),
-                self._vector_key: array_to_buffer(node.get_embedding()),
+            embedding = node.get_embedding()
+            record = {
+                NODE_ID_FIELD_NAME: node.node_id,
+                DOC_ID_FIELD_NAME: node.ref_doc_id,
+                TEXT_FIELD_NAME: node.get_content(metadata_mode=MetadataMode.NONE),
+                VECTOR_FIELD_NAME: array_to_buffer(embedding),
             }
+            # parse and append metadata
             additional_metadata = node_to_metadata_dict(
                 node, remove_text=True, flat_metadata=self.flat_metadata
             )
-            mapping.update(additional_metadata)
-
-            ids.append(node.node_id)
-            key = "_".join([self._prefix, str(node.node_id)])
-            self._redis_client.hset(key, mapping=mapping)  # type: ignore
+            data.append({**record, **additional_metadata})
```

Comment:
So, all the metadata is added (nice). I saw some examples where the custom schema specifically defines metadata fields -- what's the advantage with doing that? How is that different from the default behaviour?

Reply:
Great question. Redis keeps the "data" and "index" as separate constructs. So what you write/include in the payload here doesn't have to be added to the schema. Sometimes it's more performant or cost effective to NOT index fields you don't need to. 

There are two modes here of using Redis:

1. Default schema: in this mode, only the bare minimum, required fields will be "indexed" in Redis and searchable/filterable. However you can still store all arbitrary metadata that you want -- but it won't be included in the search index (still available to be returned along side). You can get started with this out of the box. No schema definition required.
2. Custom schema: in this mode, user can define a custom schema, which you have seen. This allows for customization of vector settings (algorithm, dims, etc) AND adding additional metadata like tags or text or numerics to the search index. Can also customize index name etc.

We could look at ways in between too if you think we need to support additional ways!

Reply:
I think that makes a ton of sense -- thank you for clarifying !